[23] NPE after PR #2935

After updating with changes from #2935 new code from that PR causes NPE:
```
Caused by: java.lang.NullPointerException: Cannot read field "id" because "expressionType" is null
	at org.eclipse.jdt.internal.compiler.ast.SwitchStatement.checkAndSetEnhanced(SwitchStatement.java:1401)
	at org.eclipse.jdt.internal.compiler.ast.SwitchStatement.resolve(SwitchStatement.java:1120)
	at org.eclipse.jdt.internal.compiler.ast.Statement.resolveWithBindings(Statement.java:496)
```

It happened with this file
```
<T extends Short> void typeVariableSingle(T single) {
    int i1 = single;
    System.out.print(i1);
    if (single instanceof int i)
        System.out.print(i);
    else
        System.out.print('-');
    switch (single) {
        case int i -> System.out.print(i);
        default -> System.out.print('-');
    }
    System.out.println();
}
<T extends Short> void typeVariableList(List<T> list) {
    int i1 = list.get(0);
    System.out.print(i1);
    if (list.get(0) instanceof int i)
        System.out.print(i);
    else
        System.out.print('-');
    switch (list.get(0)) {
        case int i -> System.out.print(i);
        default -> System.out.print('-');
    }
    System.out.println();
}
void wildcard(List<? extends Short> list) {
    int i1 = list.get(0);
    System.out.print(i1);
    if (list.get(0) instanceof int i)
        System.out.print(i);
    else
        System.out.print('-');
    switch (list.get(0)) {
        case int i -> System.out.print(i);
        default -> System.out.print('-');
    }
    System.out.println();
}
void main() {
    Short s = 1;
    typeVariableSingle(s);
    typeVariableList(Collections.singletonList(s));
    wildcard(Collections.singletonList(s));
}
```

Right now I'm seeing new resolve errors on that file: `List` cannot be found, although this should be covered by the default `import module java.base`.

@mpalat  the NPE goes to you.

@jarthana do you have an idea what broke the module import for this implicitly declared class?

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchStatement.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchStatement.java
@@ -1397,7 +1397,7 @@ public class SwitchStatement extends Expression {
 				return true;
 			}
 		}
-		if (JavaFeature.PRIMITIVES_IN_PATTERNS.isSupported(upperScope.compilerOptions())) {
+		if (expressionType != null && JavaFeature.PRIMITIVES_IN_PATTERNS.isSupported(upperScope.compilerOptions())) {
 			switch (expressionType.id) {
 				case TypeIds.T_float:
 				case TypeIds.T_double:

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PrimitiveInPatternsTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/PrimitiveInPatternsTest.java
@@ -31,7 +31,7 @@ public class PrimitiveInPatternsTest extends AbstractRegressionTest9 {
 	static {
 //		TESTS_NUMBERS = new int [] { 1 };
 //		TESTS_RANGE = new int[] { 1, -1 };
-//		TESTS_NAMES = new String[] { "testEnhancedPrimitiveSwitch_001" };
+//		TESTS_NAMES = new String[] { "testEnhancedPrimitiveSwitchNPE_001" };
 	}
 	private String extraLibPath;
 	public static Class<?> testClass() {
@@ -6933,6 +6933,30 @@ public class PrimitiveInPatternsTest extends AbstractRegressionTest9 {
 			"	switch (d) {\n" +
 			"	        ^\n" +
 			"An enhanced switch statement should be exhaustive; a default label expected\n" +
+			"----------\n");
+	}
+
+   public void testEnhancedPrimitiveSwitchNPE_001() {
+		runNegativeTest(new String[] {
+			"X.java",
+				"""
+				class X {
+					// Y is not defined/visible
+					<T extends Y> void foo(T single) {
+						switch (single) {
+						case int i -> System.out.print(i);
+						default -> System.out.print('-');
+						}
+						System.out.println("hello");
+					}
+				}
+				"""
+			},
+			"----------\n" +
+			"1. ERROR in X.java (at line 3)\n" +
+			"	<T extends Y> void foo(T single) {\n" +
+			"	           ^\n" +
+			"Y cannot be resolved to a type\n" +
 			"----------\n");
 	}
 

